features/steps/steps.py

```diff
@@ -64,36 +64,114 @@ def given_a_machine(context, series):
+                    context.instances["uaclient"].name
                 )
             )
         else:
             try:
-                context.instance.delete(wait=False)
+                context.instances["uaclient"].delete(wait=False)
             except RuntimeError as e:
                 print(
                     "Failed to delete instance: {}\n{}".format(
-                        context.instance.name, str(e)
+                        context.instances["uaclient"].name, str(e)
                     )
                 )
 
     context.add_cleanup(cleanup_instance)
-    logging.warning("--- instance ip: {}".format(context.instance.ip))
-    print("--- instance ip: {}".format(context.instance.ip))
+    logging.warning(
+        "--- instance ip: {}".format(context.instances["uaclient"].ip)
+    )
+    print("--- instance ip: {}".format(context.instances["uaclient"].ip))
+
+
+@when("I launch a `{series}` `{instance_type}` machine")
+def launch_machine(context, series, instance_type):
+    now = datetime.datetime.now()
+    date_prefix = now.strftime("-%s%f")
+    instance_name = CONTAINER_PREFIX + series + date_prefix + instance_type
+
+    context.instances[instance_type] = context.config.cloud_manager.launch(
+        series=series, instance_name=instance_name
+    )
+
+    def cleanup_instance() -> None:
+        try:
+            context.instances[instance_type].delete(wait=False)
+        except RuntimeError as e:
+            print(
+                "Failed to delete instance: {}\n{}".format(
+                    context.instances[instance_type].name, str(e)
+                )
+            )
+
+    context.add_cleanup(cleanup_instance)
+
+
+@when("I add this text on `{file_name}` on `{instance_type}` above `{line}`")
+def when_i_add_this_text_on_file_above_line(
+    context, file_name, instance_type, line
+):
+    command = 'sed -i "s/{}/{}\\n{}/" {}'.format(
+        line, context.text, line, file_name
+    )
+    when_i_run_command(
+        context, command, "with sudo", instance_type=instance_type
+    )
+
+
+@when("I run `{command}` `{user_spec}` on the `{instance_type}` machine")
+def when_i_run_command_on_machine(context, command, user_spec, instance_type):
+    when_i_run_command(
+        context, command, user_spec, instance_type=instance_type
+    )
+
+
+@when("I configure uaclient `{proxy_cfg}` using `{instance_type}` machine")
+def when_i_configure_uaclient_using_proxy_machine(
+    context, proxy_cfg, instance_type
+):
+    proxy_ip = context.instances[instance_type].ip
+
+    # We can modify this code to get this info directly from squid.conf,
+    # But I don't think we need this at this moment.
+    port = "3128"
+
+    proxy_type = "http"
+    if "https" in proxy_cfg:
+        proxy_type = "https"
+
+    proxy_cfg_entry = "{}: {}://{}:{}".format(
+        proxy_cfg, proxy_type, proxy_ip, port
+    )
+    context.text = proxy_cfg_entry
+
+    when_i_append_to_uaclient_config(context)
```

Comment:
The phrase for this command doesn't totally capture what it does. I think the phrase should explicitly say that we're configuring a proxy here. Just a slight change like this might be a little clearer:
```
When I configure uaclient `http` proxy to use `proxy` machine
```
The only real difference is that "proxy" becomes part of the phrase rather than one of the arguments, but I think it does help a bit